File: Actito.Assets.Core/Models/ActitoAsset.cs
namespace ActitoSdk.Assets.Core.Models;

/// <summary>
/// Represents a rich asset returned by Actito.
/// </summary>
/// <remarks>
/// An <see cref="ActitoAsset"/> contains displayable content such as a title,
/// optional descriptive text, a link to a binary file, and elements like a button
/// or metadata. Additional fields are stored in <see cref="ActitoAsset.extra"/>.
/// </remarks>
public class ActitoAsset
{
    /// <summary>
    ///  The title of the asset.
    /// </summary>
    public string Title { get; }

    /// <summary>
    /// Optional description of the asset.
    /// </summary>
    public string? Description { get; }

    /// <summary>
    /// Optional key of the asset.
    /// </summary>
    public string? Key { get; }

    /// <summary>
    /// Optional binary file url of the asset.
    /// </summary>
    public string? Url { get; }

    /// <summary>
    /// Optional button associated with the asset.
    /// </summary>
    public ActitoAssetButton? Button { get; }

    /// <summary>
    /// Optional metadata associated with the asset.
    /// </summary>
    public ActitoAssetMetaData? MetaData { get; }

    /// <summary>
    /// Collection of key-value pairs used to add extra information to the asset.
    /// </summary>
    public IDictionary<string, object> Extra { get; }

    /// <summary>
    /// Contructor for <see cref="ActitoAsset"/>.
    /// </summary>
    public ActitoAsset(string title, string? description, string? key, string? url, ActitoAssetButton? button,
        ActitoAssetMetaData? metaData, IDictionary<string, object> extra)
    {
        Title = title;
        Description = description;
        Key = key;
        Url = url;
        Button = button;
        MetaData = metaData;
        Extra = extra;
    }
}

/// <summary>
/// Represents a call-to-action button associated with an <see cref="ActitoAsset"/>.
/// </summary>
public class ActitoAssetButton
{
    /// <summary>
    /// Optional text displayed on the button.
    /// </summary>
    public string? Label { get; }

    /// <summary>
    /// Optional action associated with the button.
    /// </summary>
    public string? Action { get; }

    /// <summary>
    /// Constructor for <see cref="ActitoAssetButton"/>.
    /// </summary>
    public ActitoAssetButton(string? label, string? action)
    {
        Label = label;
        Action = action;
    }
}

/// <summary>
/// Contains metadata describing the underlying file of an <see cref="ActitoAsset"/>.
/// </summary>
public class ActitoAssetMetaData
{
    /// <summary>
    /// The original name of the file as provided at upload time.
    /// </summary>
    public string OriginalFileName { get; }

    /// <summary>
    /// The MIME type of the file.
    /// </summary>
    public string ContentType { get; }

    /// <summary>
    /// The size of the file in bytes.
    /// </summary>
    public int ContentLength { get; }

    /// <summary>
    /// Constructor for <see cref="ActitoAssetMetaData"/>.
    /// </summary>
    public ActitoAssetMetaData(string originalFileName, string contentType, int contentLength)
    {
        OriginalFileName = originalFileName;
        ContentType = contentType;
        ContentLength = contentLength;
    }
}

File: Actito.Assets.iOS.Binding.Bridge/Sources/Models/ActitoAsset.swift
import Foundation
import ActitoAssetsKit

@objc
public class ActitoAsset : NSObject {
    @objc public let title: String
    @objc public let assetDescription: String?
    @objc public let key: String?
    @objc public let url: String?
    @objc public let button: ActitoAssetButton?
    @objc public let metaData: ActitoAssetMetaData?
    @objc public let extra: [String: Any]

    @objc public init(title: String, assetDescription: String?, key: String?, url: String?, button: ActitoAssetButton?, metaData: ActitoAssetMetaData?, extra: [String : Any]) {
        self.title = title
        self.assetDescription = assetDescription
        self.key = key
        self.url = url
        self.button = button
        self.metaData = metaData
        self.extra = extra
    }

    public convenience init(from asset: ActitoAssetsKit.ActitoAsset) {
        self.init(
            title: asset.title,
            assetDescription: asset.description,
            key: asset.key,
            url: asset.url,
            button: asset.button.map { ActitoAssetButton(from: $0) },
            metaData: asset.metaData.map { ActitoAssetMetaData(from: $0) },
            extra: asset.extra
        )
    }
}

@objc
public class ActitoAssetButton : NSObject {
    @objc public let label: String?
    @objc public let action: String?

    @objc public init(label: String?, action: String?) {
        self.label = label
        self.action = action
    }

    public convenience init(from button: ActitoAssetsKit.ActitoAsset.Button) {
        self.init(
            label: button.label,
            action: button.action
        )
    }
}

@objc
public class ActitoAssetMetaData : NSObject {
    @objc public let originalFileName: String
    @objc public let contentType: String
    @objc public let contentLength: Int

    @objc public init(originalFileName: String, contentType: String, contentLength: Int) {
        self.originalFileName = originalFileName
        self.contentType = contentType
        self.contentLength = contentLength
    }

    public convenience init(from metaData: ActitoAssetsKit.ActitoAsset.MetaData) {
        self.init(
            originalFileName: metaData.originalFileName,
            contentType: metaData.contentType,
            contentLength: metaData.contentLength
        )
    }
}
